fastapi_sqla/async_sqla.py

```diff
@@ -108,36 +107,59 @@ async def add_session_to_request(
     """
-    async with open_session(key) as session:
-        setattr(request.state, f"{_ASYNC_REQUEST_SESSION_KEY}_{key}", session)
-        response = await call_next(request)
-
-        is_dirty = bool(session.dirty or session.deleted or session.new)
-
-        # try to commit after response, so that we can return a proper 500 response
-        # and not raise a true internal server error
-        if response.status_code < 400:
-            try:
-                await session.commit()
-            except Exception:
-                logger.exception("commit failed, returning http error")
-                response = PlainTextResponse(
-                    content="Internal Server Error", status_code=500
-                )
-
-        if response.status_code >= 400:
-            # If ever a route handler returns an http exception, we do not want the
-            # session opened by current context manager to commit anything in db.
-            if is_dirty:
-                # optimistically only log if there were uncommitted changes
-                logger.warning(
-                    "http error, rolling back possibly uncommitted changes",
-                    status_code=response.status_code,
-                )
-            # since this is no-op if session is not dirty, we can always call it
-            await session.rollback()
 
-    return response
+    def __init__(self, app: ASGIApp, key: str = _DEFAULT_SESSION_KEY) -> None:
+        self.app = app
+        self.key = key
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
+        if scope["type"] != "http":
+            return await self.app(scope, receive, send)
+
+        async with open_session(self.key) as session:
+            request = Request(scope=scope, receive=receive, send=send)
+            setattr(request.state, f"{_ASYNC_REQUEST_SESSION_KEY}_{self.key}", session)
+
+            async def send_wrapper(message: Message) -> None:
+                if message["type"] == "http.response.start":
+                    has_responded = False
+                    status_code = message["status"]
+
+                    is_dirty = bool(session.dirty or session.deleted or session.new)
+
+                    # try to commit after response, so that we can return a proper 500
+                    # and not raise a true internal server error
+                    if status_code < 400:
+                        try:
+                            await session.commit()
+                        except Exception:
+                            logger.exception("commit failed, returning http error")
+                            status_code = 500
+                            response = PlainTextResponse(
+                                content="Internal Server Error", status_code=500
+                            )
+                            await response(scope, receive, send)
+                            has_responded = True
```

Comment:
📝  Not returning directly here because we want to execute the block below with the rollback